src/test/protocol/STParsedJSON_test.cpp

```diff
@@ -35,69 +38,1963 @@ class STParsedJSON_test : public beast::unit_test::suite
+        {
+            Json::Value j;
+            j[sfEmailHash] = "01234567";
+            STParsedJSONObject obj("Test", j);
+            BEAST_EXPECT(!obj.object.has_value());
+        }
+
+        // Hex string too long for UInt128 (should fail)
+        {
+            Json::Value j;
+            j[sfEmailHash] = "0123456789ABCDEF0123456789ABCDEF00";
+            STParsedJSONObject obj("Test", j);
+            BEAST_EXPECT(!obj.object.has_value());
+        }
+
+        // Array value for UInt128 (should fail)
+        {
+            Json::Value j;
+            j[sfEmailHash] = Json::Value(Json::arrayValue);
+            STParsedJSONObject obj("Test", j);
+            BEAST_EXPECT(!obj.object.has_value());
+        }
+
+        // Object value for UInt128 (should fail)
+        {
+            Json::Value j;
+            j[sfEmailHash] = Json::Value(Json::objectValue);
+            STParsedJSONObject obj("Test", j);
+            BEAST_EXPECT(!obj.object.has_value());
+        }
+    }
+
+    void
+    testUInt160()
+    {
+        testcase("UInt160");
+        {
+            Json::Value j;
+            j[sfTakerPaysCurrency] = "0123456789ABCDEF0123456789ABCDEF01234567";
+            STParsedJSONObject obj("Test", j);
+            BEAST_EXPECT(obj.object.has_value());
+            BEAST_EXPECT(obj.object->isFieldPresent(sfTakerPaysCurrency));
+            BEAST_EXPECT(
+                obj.object->getFieldH160(sfTakerPaysCurrency).size() == 20);
+            std::array<uint8_t, 20> expected = {
+                0x01, 0x23, 0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF, 0x01, 0x23,
+                0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF, 0x01, 0x23, 0x45, 0x67};
+            BEAST_EXPECT(
+                obj.object->getFieldH160(sfTakerPaysCurrency) ==
+                uint160{expected});
+        }
+        // Valid lowercase hex string for UInt160
+        {
+            Json::Value j;
+            j[sfTakerPaysCurrency] = "0123456789abcdef0123456789abcdef01234567";
+            STParsedJSONObject obj("Test", j);
+            BEAST_EXPECT(obj.object.has_value());
+            BEAST_EXPECT(obj.object->isFieldPresent(sfTakerPaysCurrency));
+            BEAST_EXPECT(
+                obj.object->getFieldH160(sfTakerPaysCurrency).size() == 20);
+        }
+
+        // Empty string for UInt160 (should be valid, all zero)
+        {
+            Json::Value j;
+            j[sfTakerPaysCurrency] = "";
+            STParsedJSONObject obj("Test", j);
+            BEAST_EXPECT(obj.object.has_value());
+            BEAST_EXPECT(obj.object->isFieldPresent(sfTakerPaysCurrency));
+            auto const& h160 = obj.object->getFieldH160(sfTakerPaysCurrency);
+            BEAST_EXPECT(h160.size() == 20);
+            bool allZero = std::all_of(
+                h160.begin(), h160.end(), [](auto b) { return b == 0; });
+            BEAST_EXPECT(allZero);
+        }
+
+        // Odd-length hex string for UInt160 (should fail)
```

Comment:
if the size of the uintXXX values always should be const, then this test (and all the other below) is unnecessary (checked by too long and too short tests)

Reply:
e88fedfc60eb02bef99d170b22a785c8f29c2128